### PR TITLE
[ECAM]add autopilot warning

### LINKED ISSUE #1503: [REQUEST] Autopilot disconnect sound.
**Is your feature request related to a problem? Please describe.**

When disconnect autothrust, there is a no sound and no warning about disconnect autothrust

If we press autothrottle disconnect button on thurst lever's red button, a320neo should beep 'warning sound' once, and beep 3 times. 

**Describe the feature you'd like to see implemented**

**References**

https://youtu.be/Vq8Lj0pp39o?t=417
you can watch when disconnect autothurst

**Additional context**

Discord username (if different from GitHub):

### PR DESCRIPTION
**Pull request by Kimbyeoungjang**
Fixes #1503 

## Summary of Changes

add warning when disconnect autopilot, autothrottle. also add autothrthrottle disconnect sound(warning)

## References

image

https://youtu.be/g32B8_5TbdQ?t=168 <- autopilot disconnect
still need when disconnect autothrottle ecam video.

## Additional context

- [x] add warning

- [x] add warning sound

- [x] test

Discord username (if different from GitHub):

## How to download the PR for QA

Every new commit to this PR will cause a new A32NX artifact to be created and uploaded.
The build.py will have already been run with the latest changes, so no need to rerun it once you download the zip.

1. Make sure you are signed in to GitHub
1. Click on the **Checks** tab on the PR
1. On the right side, click on the **Artifacts** drop down and click the **A32NX** link

### CODE REVIEW

**Comment by Kimbyeoungjang** [comment]
i still working on this.

**Comment by lukecologne** [comment]
Then you should convert this to draft

**Comment by Kimbyeoungjang** [comment]
> Then you should convert this to draft

ok!

**Comment by lukecologne** [comment]
you closed it, that is not converting it to draft

**Comment by lukecologne** [comment]
Alright, now it is a draft

**Comment by Kimbyeoungjang** [comment]
> Alright, now it is a draft

thank you!
i'm not familiar github yet T.T

**Comment by Benjozork** [comment]
I added a technical reference to your PR, nothing to change but useful when we can't watch videos or wanna add more behavior in the future when more systems are model

**Comment by davidwalschots** [comment]
As just noted on Discord. Changing state inside the `isActive` function sounds like a bad idea, as one wouldn't expect `isActive` to have side-effects. Maybe we ought to introduce a "watcher" mechanism of sorts which can be implemented for ECAM messages that require it. Would keep the programming model cleaner.

**Comment by davidwalschots** [comment]
Note that my comment relates to the fact that `isActive` or any functions it calls shouldn't change the state. Simplify moving the state modifying logic into another function which is called by `isActive` still modifies state.

**Comment by davidwalschots** [comment]
@devsnek has some reasons why this is okay for now. See [Discord](https://discord.com/channels/738864299392630914/738864300818432023/772434672293642250). Please disregard my comments for now.

File: A32NX/html_ui/Pages/VCockpit/Instruments/Airliners/A320_Neo/EICAS/ECAM/A320_Neo_UpperECAM.js [hunk 1]
```diff
@@ -920,6 +958,22 @@ var A320_Neo_UpperECAM;
                             return (this.getCachedSimVar("L:AIRLINER_FLIGHT_PHASE", "Enum") == 6) && (this.getCachedSimVar("RADIO HEIGHT", "feet") < 800);
                         }
                     },
+                    {
+                        message: "AP OFF",
+                        style: "fail-3",
+                        important: true,
+                        isActive: () => {
+                            return (this.getAPWarning("AP", 4) == true);
```
Comment (devsnek) [inline A32NX/html_ui/Pages/VCockpit/Instruments/Airliners/A320_Neo/EICAS/ECAM/A320_Neo_UpperECAM.js]:
```js
if (this.getCachedSimVar('AUTOPILOT MASTER', 'Bool')) {
  SimVar.SetSimVarValue('L:A32NX_AP_LAST_DC_TIME', 'Number', 0);
  return false;
}
const last = SimVar.GetSimVarValue('L:A32NX_AP_LAST_DC_TIME', 'Number');
if (last > 0) {
  const now = Date.now();
  return (last - now) < 4000;
}
SimVar.SetSimVarValue('L:A32NX_AP_LAST_DC_TIME', 'Number', Date.now());
return true;
```

Comment (Kimbyeoungjang) [reply A32NX/html_ui/Pages/VCockpit/Instruments/Airliners/A320_Neo/EICAS/ECAM/A320_Neo_UpperECAM.js]: I rebuild code more simply. is it okay working on isActive?

File: A32NX/html_ui/Pages/VCockpit/Instruments/Airliners/A320_Neo/EICAS/ECAM/A320_Neo_UpperECAM.js [hunk 2]
```diff
@@ -321,6 +321,42 @@ var A320_Neo_UpperECAM;
...
+                }, count);
+                if (this.getCachedSimVar("L:ATHR_DiscWarn", "Bool") != true) {
+                    return true;
+                } else {
+                    clearTimeout(timer);
+                    return false;
+                }
+            } else {
+                SimVar.SetSimVarValue("L:ATHR_DiscWarn", "Bool", false);
+                SimVar.SetSimVarValue("L:ATHR_Disconnected", "Bool", false);
```
Comment (Benjozork) [inline A32NX/html_ui/Pages/VCockpit/Instruments/Airliners/A320_Neo/EICAS/ECAM/A320_Neo_UpperECAM.js:357]: Please prefix those LVars with `A32NX_` and put them in all capital letters

File: A32NX/html_ui/Pages/VCockpit/Instruments/Airliners/A320_Neo/EICAS/ECAM/A320_Neo_UpperECAM.js [hunk 3]
```diff
@@ -321,6 +321,42 @@ var A320_Neo_UpperECAM;
                 }
             }
         }
+        getAPWarning(_time) {
+            const count = _time * 1000;
+            let apwarning;
+            if (this.getCachedSimVar("AUTOPILOT MASTER", "Bool") == false) {
+                const timer = setTimeout(function () {
+                    SimVar.SetSimVarValue("L:AP_DiscWarn", "Bool", true);
+                }, count);
```
Comment (Benjozork) [inline A32NX/html_ui/Pages/VCockpit/Instruments/Airliners/A320_Neo/EICAS/ECAM/A320_Neo_UpperECAM.js:330]:
```suggestion
                const timer = setTimeout(() => {
                    SimVar.SetSimVarValue("L:AP_DiscWarn", "Bool", true);
                }, count);
```

File: A32NX/html_ui/Pages/VCockpit/Instruments/Airliners/A320_Neo/EICAS/ECAM/A320_Neo_UpperECAM.js [hunk 4]
```diff
@@ -321,6 +321,42 @@ var A320_Neo_UpperECAM;
...
+                SimVar.SetSimVarValue("L:AP_DiscWarn", "Bool", false);
+            }
+        }
+        getATHRWarning(_time) {
+            const count = _time * 1000;
+            if (this.getCachedSimVar("AUTOTHROTTLE ACTIVE", "Bool") == false) {
+                SimVar.SetSimVarValue("L:ATHR_Disconnected", "Bool", true);
+                const timer = setTimeout(function () {
+                    SimVar.SetSimVarValue("L:ATHR_DiscWarn", "Bool", true);
+                }, count);
```
Comment (Benjozork) [inline A32NX/html_ui/Pages/VCockpit/Instruments/Airliners/A320_Neo/EICAS/ECAM/A320_Neo_UpperECAM.js:348]:
```suggestion
                const timer = setTimeout(() => {
                    SimVar.SetSimVarValue("L:ATHR_DiscWarn", "Bool", true);
                }, count);
```

**Review by Benjozork — CHANGES_REQUESTED** [review]
LGTM. Some minor tweaks to do

**Comment by Kimbyeoungjang** [comment]
Ok i'll change that.

File: A32NX/html_ui/Pages/VCockpit/Instruments/Airliners/A320_Neo/EICAS/ECAM/A320_Neo_UpperECAM.js [hunk 5]
```diff
@@ -321,6 +321,42 @@ var A320_Neo_UpperECAM;
                 }
             }
         }
+        getAPWarning(_time) {
```
Comment (devsnek) [inline A32NX/html_ui/Pages/VCockpit/Instruments/Airliners/A320_Neo/EICAS/ECAM/A320_Neo_UpperECAM.js:324]: `_` means the variable isn't used.

```suggestion
        getAPWarning(time) {
```

File: A32NX/html_ui/Pages/VCockpit/Instruments/Airliners/A320_Neo/EICAS/ECAM/A320_Neo_UpperECAM.js [hunk 6]
```diff
@@ -321,6 +321,42 @@ var A320_Neo_UpperECAM;
...
+            const count = _time * 1000;
+            let apwarning;
+            if (this.getCachedSimVar("AUTOPILOT MASTER", "Bool") == false) {
+                const timer = setTimeout(function () {
```
Comment (devsnek) [inline A32NX/html_ui/Pages/VCockpit/Instruments/Airliners/A320_Neo/EICAS/ECAM/A320_Neo_UpperECAM.js:328]: this variable isn't persistent across invocations of `getAPWarning`. you create a timer and then clear that same timer a few lines later. this also means that if someone clicks the AP on and off a lot there will be several timers active at once.

**Review by devsnek — CHANGES_REQUESTED** [review]

File: A32NX/html_ui/Pages/VCockpit/Instruments/Airliners/A320_Neo/EICAS/ECAM/A320_Neo_UpperECAM.js (hunk 6 above)
Comment (Kimbyeoungjang) [reply A32NX/html_ui/Pages/VCockpit/Instruments/Airliners/A320_Neo/EICAS/ECAM/A320_Neo_UpperECAM.js:328]: ok i'll fix that mulitple AP disconnect situation

**Comment by Kimbyeoungjang** [comment]
i fix some simvar and funciton() to () =>
i'll fix multiple ap disconnect and add caution light.

File: A32NX/html_ui/Pages/VCockpit/Instruments/Airliners/A320_Neo/EICAS/ECAM/A320_Neo_UpperECAM.js (hunk 6 above)
Comment (Kimbyeoungjang) [reply A32NX/html_ui/Pages/VCockpit/Instruments/Airliners/A320_Neo/EICAS/ECAM/A320_Neo_UpperECAM.js]: i want to know that in JS, setTimeout still working when disconnect -> connect -> disconnect fast.
i think if i disconnect and connect fast, setTimeout is stopped working due to if. 

https://youtu.be/ZvsAmWWHnsY <- i tested the performance. i think it is okay.

should i change these code?

**Comment by Kimbyeoungjang** [comment]
@Benjozork i working on master branch. Should i change some point?

I have a question about warning system. I can beep caution sound but i don’t know how to use caution light(on captain and f/o front warning and caution light). 

Because when disconnect A/P and A/THR, caution light should turn on and turn off automatically.

Do you know how to turn on that? If that is finish, maybe i can close this pr :)

When a32nx custom a/p builded, you guys can remove my code (due to autopilot capability). At this time, i can’t add 3 deep sound(when autopilot capability reduced) because asobo don’t have that logic(i think it is displayed but there is no different XD)

**Comment by Kimbyeoungjang** [comment]
i'm finish my work. can you guys test my code?

**Comment by Kimbyeoungjang** [comment]
i find some bug, change to draft :(...

**Comment by Kimbyeoungjang** [comment]
i change for light but i worried about caution and warning light off... do you guys have a better idea for caution and warning light?

**Comment by Kimbyeoungjang** [comment]
i finished my work but please review my code.
this code include warning light but i worried about light is off.
if ap or a/thr disengaged, warning or caution light on and off but that logic can shutdown other caution or warning.
maybe delete warning light logic is best way...

**Comment by Kimbyeoungjang** [comment]
recreate pr for cleanup.

**Pull request closed without merging**